### PR TITLE
Move repo to newer version of Public Api analyzers

### PR DESCRIPTION
**Pull request by mavasani**

### CODE REVIEW

**Comment by mavasani** [comment]
@jasonmalinowski It seems the correctness analyzers leg is failing as you recently added a public API that has no entry in unshipped file:
```
src/Workspaces/Core/Portable/Workspace/ProjectSystem/IFileChangeWatcher.cs(26,25): error RS0016: (NETCORE_ENGINEERING_TELEMETRY=Build) Symbol 'WatchedDirectory' is not part of the declared public API
```

Looking at your PR which added this public API: https://github.com/dotnet/roslyn/pull/65701, it seems the public API addition was likely unintentional. I am going to to switch it to an internal type, and feel free to make it public again if that was an intentional change.

**Review by 333fred — APPROVED** [review]

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 3)

**Comment by mavasani** [comment]
Filed #65746 to review all the non-infra changes in this PR. I am going to merge this PR as soon as it is green to unblock re-enabling the public API analyzer and avoid more public API changes to unintentionally sneak in.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 1]
```diff
@@ -3331,7 +3331,7 @@ public override SyntaxNode BaseExpression()
         public override SyntaxNode TypedConstantExpression(TypedConstant value)
             => ExpressionGenerator.GenerateExpression(value);
 
-        protected override SyntaxNode GenerateExpression(ITypeSymbol? type, object? value, bool canUseFieldReference)
+        private protected override SyntaxNode GenerateExpression(ITypeSymbol? type, object? value, bool canUseFieldReference)
```
Comment (mavasani) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:3334]: I don't believe this was intended to be a public API, but should be confirmed with Cyrus as part of #65746

File: src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -15,6 +15,8 @@ Microsoft.CodeAnalysis.Editing.DeclarationModifiers.IsFile.get -> bool
 Microsoft.CodeAnalysis.Editing.DeclarationModifiers.WithIsFile(bool isFile) -> Microsoft.CodeAnalysis.Editing.DeclarationModifiers
 *REMOVED*static Microsoft.CodeAnalysis.Editing.SyntaxGenerator.DefaultRemoveOptions -> Microsoft.CodeAnalysis.SyntaxRemoveOptions
 Microsoft.CodeAnalysis.Editing.SyntaxEditor.SyntaxEditor(Microsoft.CodeAnalysis.SyntaxNode root, Microsoft.CodeAnalysis.Host.SolutionServices services) -> void
+*REMOVED*abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.LiteralExpression(object value) -> Microsoft.CodeAnalysis.SyntaxNode
+Microsoft.CodeAnalysis.Editing.SyntaxGenerator.LiteralExpression(object value) -> Microsoft.CodeAnalysis.SyntaxNode
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt:19]: These 2 changes should be confirmed with Cyrus as part of #65746. It seems he converted an existing public abstract method to a public non-abstract method.

File: src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -52,8 +54,7 @@ override Microsoft.CodeAnalysis.LoadTextOptions.GetHashCode() -> int
 override Microsoft.CodeAnalysis.LoadTextOptions.ToString() -> string
 static Microsoft.CodeAnalysis.LoadTextOptions.operator !=(Microsoft.CodeAnalysis.LoadTextOptions left, Microsoft.CodeAnalysis.LoadTextOptions right) -> bool
 static Microsoft.CodeAnalysis.LoadTextOptions.operator ==(Microsoft.CodeAnalysis.LoadTextOptions left, Microsoft.CodeAnalysis.LoadTextOptions right) -> bool
-virtual Microsoft.CodeAnalysis.FileTextLoader.CreateText(System.IO.Stream stream, Microsoft.CodeAnalysis.LoadTextOptions options, System.Threading.CancellationToken cancellationToken) -> Microsoft.CodeAnalysis.Text.SourceText
-*REMOVED*override Microsoft.CodeAnalysis.FileTextLoader.LoadTextAndVersionAsync(Microsoft.CodeAnalysis.Workspace workspace, Microsoft.CodeAnalysis.DocumentId documentId, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<Microsoft.CodeAnalysis.TextAndVersion>
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt:56]: TextLoader and FileTextLoader related changes in this file should be reviewed with Tomas as part of #65746

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/IFileChangeWatcher.cs [hunk 4]
```diff
@@ -23,7 +23,7 @@ internal interface IFileChangeWatcher
     /// just creating a single directory watch on the root of a project for source file changes: rather than creating a file watcher
     /// for each individual file, we can just watch the entire directory and that's it.
     /// </remarks>
-    public sealed class WatchedDirectory
+    internal sealed class WatchedDirectory
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/IFileChangeWatcher.cs:26]: Changes to this file should be confirmed with JasonMal as part of #65746

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/IFileChangeWatcher.cs:26]: Ah yes, that wasn't intended to be made public, thanks for catching that!

File: src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt:19]: yes.  that's correct (and safe from an ABI perspective).  Thsi type cannot be publicly overridden either, so this is fine.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:3334]: SyntaxGenerator is not publicly inheritable (it has tons of internal abstract members).  As such, all 'protected' members are save to change in any fashion.

Note: i would love it public-api-analyzer understood this.  I think it understands if there is an explicit internal constructor, but not if there are internal abstracts.

**Pull request merged**